File: js/src/StringProblem/TextJustification.js



export const Justify=(words , maxWidth)=>{

	let currentLineLength = 0;

	let currentLineWords = [];

	let result = [];

	for (let i=0; i < words.length; i++) {
		
		if(currentLineLength + currentLineWords.length + words[i].length > maxWidth){

			let maxSpaceAvaliable = maxWidth - currentLineLength;

			if(currentLineWords.length === 1){

				result.push(currentLineWords[0] + " ".repeat(maxSpaceAvaliable));
	
			}else{

				let spaceNeeded = currentLineWords.length - 1;

				let evenlySpace = Math.floor(maxSpaceAvaliable / spaceNeeded);

				let spaceExtra = maxSpaceAvaliable % spaceNeeded;

				let justifyLine = "";

				for(let j=0; j < spaceNeeded; j++){

					justifyLine += currentLineWords[j] + " ".repeat(evenlySpace + ( j < spaceExtra? 1:0) );
				}

				justifyLine += currentLineWords[currentLineWords.length -1];

				result.push(justifyLine);

			}

			currentLineLength = 0;

			currentLineWords = [];
		}

		currentLineWords.push(words[i]);

		currentLineLength += words[i].length;

	}
		let JustifyLeftLastLine = currentLineWords.join(" ");

		result.push(JustifyLeftLastLine + " ".repeat(maxWidth - JustifyLeftLastLine.length));

		return result;

}

File: python/string_problem/text_justification.py
import math

def justify(words, maxWidth):
    currentLineLength = 0
    currentLineWords = []
    result = []

    for word in words:
        if currentLineLength + len(currentLineWords) + len(word) > maxWidth:
            spaceAvailable = maxWidth - currentLineLength

            if len(currentLineWords) == 1:
                result.append(currentLineWords[0] + " " * spaceAvailable)
            else:
                spaceNeeded = len(currentLineWords) - 1
                evenlySpace = math.floor(spaceAvailable / spaceNeeded)
                extraSpace = spaceAvailable % spaceNeeded

                justifyLine = ""
                for i in range(spaceNeeded):
                    justifyLine += currentLineWords[i] + (" " * (evenlySpace + (1 if i < extraSpace else 0)))

                justifyLine += currentLineWords[len(currentLineWords) - 1]
                result.append(justifyLine)

            currentLineWords = []
            currentLineLength = 0

        currentLineLength += len(word)
        currentLineWords.append(word)

    # Handle the last line (left-aligned)
    lastLine = " ".join(currentLineWords)
    result.append(lastLine + (" " * (maxWidth - len(lastLine))))

    return result
